### DISCUSSION TITLE
Guidance on how to compile w/ support for A-axis pins as Spindle

### DISCUSSION THREAD

**Discussion by johnobrien** [grblHAL/RP2040#90] [Q&A]
Greeting!

I have a PicoCNC setup and working. I am attempting to use a G320X servo drive to drive a servo as a spindle. What I'd like to do is use the A-Axis pins to send STEP and DIR signals to the G320X whenever the relevant M03 S100 G-code is executed.

I can see in the spindle plugin that there is a file called stepper.c (https://github.com/grblHAL/Plugins_spindle/blob/master/stepper.c) which seems to do what I need.

I have VS Code setup and can compile RP2040 for the PicoCNC. What I don't know how to do is 1) set the correct options in my_machine.h to build properly for PicoCNC and 2) how to "include" and "enable" the spindle plugin and the stepper.c option to be able to then use the A-Axis to drive the G320X to drive the servo and use it as a spindle.

I have attempted to search similar discussions but not found anything; apologies if the answer has already been provided and I just missed it.

Can anyone help me out?

Regards,
John

**Comment by terjeio** [comment]
To enable uncomment [this line](https://github.com/grblHAL/RP2040/blob/339e6450c965c3b0e2c6d41e30fe74b7c5ed75ac/my_machine.h#L48) and change it to: `#define SPINDLE0_ENABLE  SPINDLE_STEPPER`

**Comment by johnobrien** [comment]
Thank you terjeio. When I build the grblHAL.uf2 file and flash it to the pi pico, I cannot communicate with the pi pico via IO sender. Please see screenshot below for the error message I receive.
![Screenshot 2024-06-06 115529](https://github.com/grblHAL/RP2040/assets/446719/7dac169a-29e5-4c1a-ad25-683bf5cfb079)

I suspect that I am not editing the my_machine.h file correctly before building the grbhHAL.uf2 file. Previously I have built the firmware using the web firmware tool, so I do not know the correct edits to make to the my_machine.h file to correctly configure it to utilize the PicoCNC board (ideally with ethernet enabled). Can any either provide a correct my_machine.h file, point me to an example of a correctly edited my_machine.h file to use, or suggest the proper edits I should make to my my_machine.h file?

Contents of my my_machine.h file copy/pasted below.

`/*
  my_machine.h - configuration for Raspberry RP2040 ARM processors

  Part of grblHAL

  Copyright (c) 2021-2024 Terje Io

  grblHAL is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  grblHAL is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with grblHAL. If not, see <http://www.gnu.org/licenses/>.
*/

// NOTE: Only one board may be enabled!
// If none is enabled pin mappings from generic_map.h will be used.
#define BOARD_PICO_CNC
//#define BOARD_PICOBOB
//#define BOARD_PICOBOB_G540
//#define BOARD_PICOBOB_DLX
//#define BOARD_PICOBOB_DLX_G540
//#define BOARD_BTT_SKR_PICO_10 // incomplete and untested!
//#define BOARD_CNC_BOOSTERPACK
//#define BOARD_CITOH_CX6000    // C.ITOH CX-6000 HPGL plotter
//#define BOARD_GENERIC_4AXIS
//#define BOARD_MY_MACHINE      // Add my_machine_map.h before enabling this!

// Configuration
// Uncomment to enable.

#ifndef USB_SERIAL_CDC
#define USB_SERIAL_CDC          1 // Serial communication via native USB.
#endif
//#define BLUETOOTH_ENABLE        2 // Set to 2 for HC-05 module, enable in CMakeLists.txt if for Pico W Bluetooth.
// Spindle selection:
// Up to four specific spindle drivers can be instantiated at a time
// depending on N_SPINDLE and N_SYS_SPINDLE definitions in grbl/config.h.
// If none are specified the default PWM spindle is instantiated.
// Spindle definitions can be found in grbl/spindle_control.h.
// More here https://github.com/grblHAL/Plugins_spindle
#define SPINDLE0_ENABLE SPINDLE_STEPPER
//#define SPINDLE1_ENABLE         SPINDLE_PWM0
//#define SPINDLE2_ENABLE         SPINDLE_NONE 
//#define SPINDLE2_ENABLE         SPINDLE_NONE
// **********************
//#define MODBUS_ENABLE           1 // Set to 1 for auto direction, 2 for direction signal on auxillary output pin.
//#define WIFI_ENABLE             0 // Do NOT enable here, enable in CMakeLists.txt!
//#define WIFI_SOFTAP             1 // Use Soft AP mode for WiFi. NOTE: WIP - not yet complete!
//#define ETHERNET_ENABLE         0 // Do NOT enable here, enable in CMakeLists.txt!
//#define _WIZCHIP_            5500 // Selects WIZnet ethernet breakout connected via SPI.
                                    // Uncomment to enable W5500 chip, default is W5100S. Requires ethernet enabled in CMakeLists.txt.
//#define WEBUI_ENABLE            1 // Enable ESP3D-WEBUI plugin along with networking and SD card plugins. Requires WiFi enabled.
//#define WEBUI_AUTH_ENABLE       1 // Enable ESP3D-WEBUI authentication.
//#define WEBUI_INFLASH           1 // Store WebUI files in flash instead of on SD card.
//#define SDCARD_ENABLE           2 // Run gcode programs from SD card. Set to 2 to enable YModem upload.
//#define MPG_ENABLE              1 // Enable MPG interface. Requires a serial port and means to switch between normal and MPG mode.
                                    // 1: Mode switching is by handshake pin input unless the keypad plugin is enabled in mode 2 which
                                    //    uses mode switching by the CMD_MPG_MODE_TOGGLE (0x8B) command character.
                                    // 2: Mode switching is by the CMD_MPG_MODE_TOGGLE command character. The keypad plugin is not required.
//#define KEYPAD_ENABLE           1 // 1: uses a I2C keypad for input.
                                    // 2: uses a serial port for input. If MPG_ENABLE is set to 1 the serial stream is shared with the MPG.
//#define DISPLAY_ENABLE          1 // Set to 1 for I2C display protocol, 2 for I2C LED protocol.
//#define ODOMETER_ENABLE         1 // Odometer plugin.
//#define LASER_COOLANT_ENABLE    1 // Laser coolant plugin. To be completed.
//#define LB_CLUSTERS_ENABLE      1 // LaserBurn cluster support.
//#define FANS_ENABLE             1 // Enable fan control via M106/M107. Activates fan plugin.
//#define EMBROIDERY_ENABLE       1 // Embroidery plugin. To be completed.
//#define TRINAMIC_ENABLE         1 // Trinamic TMC2130 stepper driver support. NOTE: work in progress.
//#define TRINAMIC_I2C            1 // Trinamic I2C - SPI bridge interface.
//#define TRINAMIC_DEV            1 // Development mode, adds a few M-codes to aid debugging. Do not enable in production code.
//#define EEPROM_ENABLE          16 // I2C EEPROM/FRAM support. Set to 16 for 2K, 32 for 4K, 64 for 8K, 128 for 16K and 256 for 16K capacity.
//#define EEPROM_IS_FRAM          1 // Uncomment when EEPROM is enabled and chip is FRAM, this to remove write delay.
//#define STEP_INJECT_ENABLE      1
// Optional control signals:
// These will be assigned to aux input pins. Use the $pins command to check which pins are assigned.
// NOTE: If not enough pins are available assignment will silently fail.
//#define PROBE_ENABLE            0 // Default enabled, remove comment to disable probe input.
//#define SAFETY_DOOR_ENABLE      1
//#define MOTOR_FAULT_ENABLE      1
//#define MOTOR_WARNING_ENABLE    1
//#define PROBE_DISCONNECT_ENABLE 1
//#define STOP_DISABLE_ENABLE     1
//#define BLOCK_DELETE_ENABLE     1
//#define SINGLE_BLOCK_ENABLE     1
//#define LIMITS_OVERRIDE_ENABLE  1

// If the selected board map supports more than three motors ganging and/or auto-squaring
// of axes can be enabled here.
//#define X_GANGED             1
//#define X_AUTO_SQUARE        1
//#define Y_GANGED             1
//#define Y_AUTO_SQUARE        1
//#define Z_GANGED             1
//#define Z_AUTO_SQUARE        1
// For ganged axes the limit switch input (if available) can be configured to act as a max travel limit switch.
// NOTE: If board map already has max limit inputs defined this configuration will be ignored.
//#define X_GANGED_LIM_MAX     1
//#define Y_GANGED_LIM_MAX     1
//#define Z_GANGED_LIM_MAX     1
//

//#define _WIZCHIP_ 5500

#if WIFI_ENABLE || ETHERNET_ENABLE || WEBUI_ENABLE
#define TELNET_ENABLE        1 // Telnet daemon - requires WiFi streaming enabled.
#define WEBSOCKET_ENABLE     1 // Websocket daemon - requires WiFi streaming enabled.
//#define MDNS_ENABLE          0 // mDNS daemon. Do NOT enable here, enable in CMakeLists.txt!
//#define SSDP_ENABLE          1 // SSDP daemon - requires HTTP enabled.
#if SDCARD_ENABLE || WEBUI_ENABLE
#define FTP_ENABLE           1 // Ftp daemon - requires SD card enabled.
//#define HTTP_ENABLE          1 // http daemon - requires SD card enabled.
//#define WEBDAV_ENABLE        1 // webdav protocol - requires http daemon and SD card enabled.
#endif
// The following symbols have the default values as shown, uncomment and change as needed.
//#define NETWORK_STA_HOSTNAME    "grblHAL"
//#define NETWORK_STA_IPMODE      1 // 0 = static, 1 = DHCP, 2 = AutoIP
//#define NETWORK_STA_IP          "192.168.5.1"
//#define NETWORK_STA_GATEWAY     "192.168.5.1"
//#define NETWORK_STA_MASK        "255.255.255.0"
#if WIFI_SOFTAP > 0
//#define NETWORK_AP_SSID         "grblHAL_AP"
//#define NETWORK_AP_PASSWORD     "grblHALap"
//#define NETWORK_AP_HOSTNAME     "grblHAL_AP"
//#define NETWORK_AP_IPMODE       0              // Do not change!
//#define NETWORK_AP_IP           "192.168.4.1"  // Do not change!
//#define NETWORK_AP_GATEWAY      "192.168.4.1"  // Do not change!
//#define NETWORK_AP_MASK         "255.255.255.0"
#endif
//#define NETWORK_FTP_PORT     21
//#define NETWORK_TELNET_PORT  23
//#define NETWORK_HTTP_PORT    80
#if HTTP_ENABLE
//#define NETWORK_WEBSOCKET_PORT  81
#else
//#define NETWORK_WEBSOCKET_PORT  80
#endif
#endif // WIFI_ENABLE

/**/
`

**Reply by terjeio** [reply]
Your my_machine.h looks ok, did you change [grbl/config.h](https://github.com/grblHAL/core/blob/0a86e57ddc622c758a2c490f4100af948beebee2/config.h#L42) to enable the A-axis?

BTW here is the firmware compiled with ethernet, A-axis and stepper spindle enabled:
[grblHAL.zip](https://github.com/user-attachments/files/15688556/grblHAL.zip)

**Comment by johnobrien** [comment]
Thank you so much! The firmware you sent along works perfectly!

No, I did not change grbl/config.h to enable the A-axis. How do I do that?

In addition, did you make any other changes to any files to compile the firmware that you sent me? When I make the changes to my my_machine.h and CMakeLists.txt to match yours, then compile and flash the resulting firmware to the Pico, I get the same error message I posted previously, which leads me to think 1) I am perhaps using a different version of the RP2040 repo or something than what you are using or 2) there were other changes you perhaps made in addition to the changes to the files you sent me. Could either of those things be possible?

I ask because eventually I want to look into getting spindle syncing to work, so if I can get to a state where I can compile the firmware myself, that would be ideal, so I could them modify from that known-good state to some new state to try to get spindle syncing working.

**Reply by terjeio** [reply]
> No, I did not change grbl/config.h to enable the A-axis. How do I do that?

By changing [this define](https://github.com/grblHAL/core/blob/0a86e57ddc622c758a2c490f4100af948beebee2/config.h#L42) to 4.

>  I am perhaps using a different version of the RP2040 repo

Check your [version](https://github.com/grblHAL/core/blob/0a86e57ddc622c758a2c490f4100af948beebee2/grbl.h#L45) (build date) against the copy you have.

**Comment by johnobrien** [comment]
My version matches the version/build date you indicated. I have tried building it using Platform IO in Visual Studio Code and Raspberry Pi Pico extension in Visual Studio Code. Neither of the resulting firmware's work when flashed to the Pico.

It is probably something I have done wrong on my end that is outside the scope of the code in the repo, so, what I figure I'll do is thank you again (thanks again!) for your help, and close this ticket for now, as the firmware you provided works.
